aclk/agent_cloud_link.c

```diff
@@ -131,6 +133,57 @@ int cloud_to_agent_parse(JSON_ENTRY *e)
 }
 
 
+static RSA *aclk_private_key = NULL;
+static int create_private_key()
+{
+    char filename[FILENAME_MAX + 1];    struct stat statbuf;
+    snprintfz(filename, FILENAME_MAX, "%s/claim.d/private.pem", netdata_configured_user_config_dir);
+
+    if (lstat(filename, &statbuf) != 0) {
+        error("Claimed agent cannot establish ACLK - private key not found '%s' failed reason=\"%s\".", filename, strerror(errno));
+        return 1;
+    }
+    if (unlikely(statbuf.st_size == 0)) {
+        info("Claimed agent cannot establish ACLK - private key '%s' is empty.", filename);
+        return 1;
+    }
+
+    FILE *f = fopen(filename, "rt");
+    if (unlikely(f == NULL)) {
+        error("Claimed agent cannot establish ACLK - unable to open private key '%s' failed reason=\"%s\".", filename, strerror(errno));
+        return 1;
+    }
+
+    char *private_key = callocz(1, statbuf.st_size + 1);
+    size_t bytes_read = fread(private_key, 1, statbuf.st_size, f);
+    private_key[bytes_read] = 0;
+    debug(D_ACLK, "Claimed agent loaded private key len=%zu bytes", bytes_read);
+    fclose(f);
+
+    BIO *key_bio = BIO_new_mem_buf(private_key, -1);
+    if (key_bio==NULL) {
+        error("Claimed agent cannot establish ACLK - failed to create BIO for key");
+        goto biofailed;
+    }
+
+    aclk_private_key = PEM_read_bio_RSAPrivateKey(key_bio, NULL, NULL, NULL);
+    BIO_free(key_bio);
+    if (aclk_private_key!=NULL)
+    {
+        freez(private_key);
+        return 0;
+    }
+    char * err = mallocz(130);;
+    ERR_load_crypto_strings();
+    ERR_error_string(ERR_get_error(), err);
```

Comment:
From the manpage "buf must be at least 256 bytes long.", this can cause memory corruption. Maybe use `ERR_error_string_n()` instead or remove the magic number 130?

Reply:
Yikes - I copied that from `security.c` so we have the same problem on the TLS in the web server.

Reply:
Will fix.

Reply:
I've misremember the code in `security.c` - it implements the check in the way that you suggest. I'll update this to do the same.